### PR TITLE
Fixed a bug in computing the langevin torques applied to rigid bodies

### LINKED ISSUE #2288: [BUG] fix rigid/nve langevin not working as expected
**Summary**

The Langevin thermostat is not working as expected in conjunction with fix rigid/nve and rigid/nve/small.

**LAMMPS Version and Platform**

LAMMPS version: 3 Mar 2020
Running on Ubuntu 16.04.7.

**Expected Behavior**

I am running an NVT simulation of rigid molecules. Each molecule is composed by three particles, one of type 1 and two of type 2. The type 2 particles are arranged in a certain configuration at fixed distance 0.5 from the center of the type 1 particle. I am using fix rigid/nve/small langevin. Intramolecular interactions are disabled (neigh_modify exclude molecule/intra all). See input file reported below for details. 

The temperature is expected to fluctuate around the target value (T=0.05).

**Actual Behavior**

An explosion of the kinetic energy is observed when using as input the file 1.lammpsdata.
Using a _very similar_ configuration (see attached file 2.lammpsdata), in which the three particles in a molecule are arranged on the same line, gives reasonable behavior. Note that in 1.lammpsdata the three particles in a molecule are "almost" arranged on the same line, but one out of three is slightly misaligned.

- Reducing the time step does not seem to solve the problem.

- Changing the interaction potentials (1-1, 1-2 and 2-2) does not seem to solve the problem. I have tried different combinations, for example:

```
pair_style hybrid lj/cut 1.122462048 zero 1.0
pair_coeff 1 1 lj/cut 1.0 1.0 1.122462048
pair_modify pair lj/cut shift yes
pair_coeff 1 2 zero 1.122462048
pair_coeff 2 2 zero 1.122462048

pair_style lj/cut 1.122462048
pair_modify shift yes
pair_coeff 1 1 1.0 1.0
pair_coeff 1 2 0.0 1.0 1.0
pair_coeff 2 2 0.0 1.0 1.0

pair_style lj/cut 1.122462048
pair_modify shift yes
pair_coeff * * 1.0 1.0 1.122462048
```

, getting always the same result.

- Visual inspection of the trajectory (using VMD) reveals that when using the file 1.lammpsdata some particles start to rotate very rapidly, resulting in the observed explosion of the kinetic energy (it may be necessary to generate many dump files, however a very short run is sufficient).

- Using fix rigid/nve instead of rigid/nve/small gives the same behavior.

- Using fix rigid/nvt and rigid/nvt/small gives **reasonable** behavior.

**Steps to Reproduce**

I attach two input files:

- 1.lammpsdata --> anomalous behavior

- 2.lammpsdata --> "reasonable" behavior

[data_files.zip](https://github.com/lammps/lammps/files/5069572/data_files.zip)

The input script which I am using is the following (commented lines are different pair styles that may be tried):

```
variable nsteps equal 1e6
variable mytemp         equal 0.05    
variable thermodump     equal   $(v_nsteps/1e5)
variable trajdump       equal   $(floor(v_nsteps/20))    
variable trestart       equal   $(floor(v_nsteps/5.))    
variable tstep         equal   0.001  
variable damp   equal   0.1    
variable neigh_cutoff equal 0.3
variable lseed equal 921389
variable vseed equal 191814
variable fname string data
variable thermofile     string  thermo.dat
variable simname string run_T${mytemp}

units           lj          
boundary        p p p      
atom_style     bond

read_data       ${fname}

neighbor ${neigh_cutoff} bin
neigh_modify every 10 delay 0 check yes
neigh_modify exclude molecule/intra all

#pair_style hybrid lj/cut 1.122462048 zero 1.0
#pair_coeff 1 1 lj/cut 1.0 1.0 1.122462048
#pair_modify pair lj/cut shift yes
#pair_coeff 1 2 zero 1.122462048
#pair_coeff 2 2 zero 1.122462048

#pair_style lj/cut 1.122462048
#pair_modify shift yes
#pair_coeff 1 1 1.0 1.0
#pair_coeff 1 2 0.0 1.0 1.0
#pair_coeff 2 2 0.0 1.0 1.0

pair_style lj/cut 1.122462048
pair_modify shift yes
pair_coeff * * 1.0 1.0 1.122462048

timestep ${tstep}

variable step equal step
variable temp equal temp
variable etot equal etotal
variable ke equal ke
variable pe equal pe
variable press equal press

thermo_style custom step etotal ke pe temp press
thermo ${thermodump}
thermo_modify norm no

fix fix_rigid all rigid/nve/small molecule langevin ${mytemp} ${mytemp} ${damp} ${lseed}

velocity all create ${mytemp} ${vseed} dist gaussian mom yes rot yes
run 0
velocity all scale ${mytemp}

fix fix_print all print ${thermodump} "${step} ${etot} ${ke} ${pe} ${temp} ${press}" file ${thermofile} screen no title "#1:step 2:etot 3:ke 4:pe 5:temp 6:press"

dump 1 all custom ${trajdump} dump/dump.${simname}.*.lammpstrj type mol xu yu zu
dump_modify 1 sort id
dump_modify 1 pad 10
dump_modify 1 format line "%d %d %.8f %.8f %.8f"

run ${nsteps}

unfix fix_print
unfix fix_rigid

write_data      data.${simname}
write_restart restart/restart.${simname}
```

### PR DESCRIPTION
**Pull request by ndtrung81**
**Summary**

This PR addresses the issue reported by #2288 where the langevin thermostat inside fix rigid and fix rigid/small (and their variants) does not function as expected.

**Related Issues**

Closes #2288 

**Author(s)**

Trung Nguyen (Northwestern)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

The PR should respect backward compatibility.

**Implementation Notes**

From the tests done by @valesori regarding the issue #2288, it seems that the body rotational degrees of freedom are not thermostated properly. This is because the langevin torques (langextra[3], langextra[4] and langextra[5]) to be added to the bodies should be in the space frame. Originally these torques were computed from the principal moments of inertia (defined in the body frame) and the body angular velocities in the space frame (omega), which are inconsistent with each other. The changes in this PR do the following: 1) converting the body angular velocities from the space frame into the body frame, then 2) computing the langevin torques in the body frame, and 3) converting the torques from the body frame back to the space frame. 

Since other fix rigid variants derive from FixRigid or from FixRigidSmall, the changes in this PR are applicable to those variants as well.

Currently, the temporary variables wbody[3] and tbody[3] and the convertion back and forth between the space and body frames are for code readability.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

### CODE REVIEW

File: src/RIGID/fix_rigid.cpp [hunk 1]
```diff
@@ -991,12 +991,33 @@ void FixRigid::apply_langevin_thermostat()
 
       gamma1 = -1.0 / t_period / ftm2v;
       gamma2 = tsqrt * sqrt(24.0*boltz/t_period/dt/mvv2e) / ftm2v;
-      langextra[i][3] = inertia[i][0]*gamma1*omega[i][0] +
+
+      // convert omega from space frame to body frame
+
```
Comment (sjplimp) [inline src/RIGID/fix_rigid.cpp:996]: Is this the same as:
MathExtra::matvec(ex_space[i],ey_space[i],ez_space[i],omega[i],wbody);
and ditto below for:
MathExtra::matvec(ex_space[i],ey_space[i],ez_space[i],tbody,&langextra[3]);
I.e. replace 3 lines with 1. ??

**Review by sjplimp — CHANGES_REQUESTED** [review]
See my comment in code ...

File: src/RIGID/fix_rigid.cpp (hunk 1 above)
Comment (ndtrung81) [reply src/RIGID/fix_rigid.cpp:996]: @sjplimp thanks for pointing to the existing helper functions, the first is transpose_matvec(), the latter is matvec()

MathExtra::transpose_matvec(ex_space[i],ey_space[i],ez_space[i],omega[i],wbody);
MathExtra::matvec(ex_space[i],ey_space[i],ez_space[i],tbody,&langextra[i][3]);

I will make the changes as requested.

**Comment by sjplimp** [comment]
Thanks for fixing this issue @ndtrung81

**Review by sjplimp — APPROVED** [review]
In my earlier comment, I meant thanks Trung for fixing the Langevin issue raised by the user.

**Review by akohlmey — APPROVED** [review]

**Pull request merged**